nimbus/src/enrollment.rs

```diff
@@ -20,28 +23,300 @@ pub enum EnrolledReason {
+        is_user_participating: bool,
+        nimbus_id: &Uuid,
+        available_randomization_units: &AvailableRandomizationUnits,
+        app_context: &AppContext,
+        updated_experiment: &Experiment,
+        enrollment_events: &mut Vec<EnrollmentChangeEvent>,
+    ) -> Result<Option<Self>> {
+        Ok(Some(match &self.status {
+            EnrollmentStatus::NotEnrolled { .. } => {
+                if !is_user_participating || updated_experiment.is_enrollment_paused {
+                    self.clone()
+                } else {
+                    let updated_enrollment = evaluate_enrollment(
+                        nimbus_id,
+                        available_randomization_units,
+                        app_context,
+                        updated_experiment,
+                    )?;
+                    log::debug!(
+                        "Experiment '{}' with enrollment {:?} is now {:?}",
+                        &self.slug,
+                        &self,
+                        updated_enrollment
+                    );
+                    if let EnrollmentStatus::Enrolled {
+                        branch,
+                        enrollment_id,
+                        ..
+                    } = &updated_enrollment.status
+                    {
+                        enrollment_events.push(EnrollmentChangeEvent::new(
+                            &updated_enrollment.slug,
+                            enrollment_id,
+                            branch,
+                            None,
+                            EnrollmentChangeEventType::Enrollment,
+                        ));
+                    };
+                    updated_enrollment
+                }
+            }
+            EnrollmentStatus::Enrolled {
+                branch,
+                enrollment_id,
+                ..
+            } => {
+                let branch = branch.to_owned();
+                let enrollment_id = enrollment_id.to_owned();
+                if !is_user_participating {
+                    log::debug!(
+                        "Existing experiment enrollment '{}' is now disqualified (global opt-out)",
+                        &self.slug
+                    );
+                    enrollment_events.push(EnrollmentChangeEvent::new(
+                        &self.slug,
+                        &enrollment_id,
+                        &branch,
+                        Some("optout"),
+                        EnrollmentChangeEventType::Disqualification,
+                    ));
+                    Self {
+                        slug: self.slug.clone(),
+                        status: EnrollmentStatus::Disqualified {
+                            reason: DisqualifiedReason::OptOut,
+                            enrollment_id,
+                            branch,
+                        },
+                    }
+                } else {
+                    let mut updated_enrollment = evaluate_enrollment(
```

Comment:
Reading Ryan's document, we need to handle at least the fact that the experiment may now be closed for new enrollments, but we still need to stay enrolled. It's not clear to me from that doc what should happen if targeting changed, but I *think* the intent is that the original enrollment remains in that case too?